marshal/admin.go

```diff
@@ -0,0 +1,322 @@
+// release releases claim on topic, partition. Optionally resets the offset on the partition.
+func (a consumerGroupAdmin) release(c claimAttempt, withNewOffset bool) bool {
+	releaseOffset := c.currentOffset
+	if withNewOffset {
+		if err := a.marshaler.Heartbeat(c.topic, c.partID, c.newOffset); err != nil {
+			log.Errorf("[%s:%d] Admin failed to heartbeat", c.topic, c.partID)
+			return false
+		}
+		releaseOffset = c.newOffset
+	}
+
+	if err := a.marshaler.ReleasePartition(c.topic, c.partID, releaseOffset); err != nil {
+		log.Errorf("[%s:%d] Admin failed to release partition with offset %d: %s",
+			c.topic, c.partID, releaseOffset, err)
+		return false
+	}
+	return true
+}
+
+// releaseClaims releases all claims the Admin has, optionally resetting their offsets.
+func (a *consumerGroupAdmin) releaseClaims(newOffset bool) error {
+	a.lock.RLock()
+	defer a.lock.RUnlock()
+
+	if !newOffset {
+		log.Infof("Admin releasing claims without resetting offsets.")
+	}
+
+	results := make(chan bool)
+	defer close(results)
+	for _, claimAttempt := range a.claims {
+
+		// Update our step counter. Releasing with a new offset will Heartbeat then ReleasePartition,
+		// releasing without resetting the offset will just ReleasePartition
+		atomic.AddInt32(a.steps, 1)
+		if newOffset {
+			atomic.AddInt32(a.steps, 1)
+		}
+
+		go func() {
+			results <- a.release(claimAttempt, newOffset)
+		}()
+	}
+
+	numResults := len(a.claims)
+	for numResults > 0 {
+		select {
+		case result := <-results:
+			if !result {
+				return fmt.Errorf("Admin failed to set a Kafka offset. Aborting!")
+			}
+			numResults--
+		default:
+		}
+	}
+	return nil
+}
+
+// heartbeatLoop hearbeats as if we had a claim to this partition and were simply
+// not reading past where the previous owner had left off.
+func (a *consumerGroupAdmin) heartbeatLoop(
+	topic string, partID int, lastOffset int64, stopChan chan struct{}) {
+
+	for {
+		select {
+		// Stop claimHealth either when all topic, partitions have been successfully claimed,
+		// or the Admin has failed to do so and needs to abort.
+		case <-stopChan:
+			return
+		default:
+			// If we fail to heartbeat, record this in claimHealth.
+			// The Admin will take care of cleaning up other claims.
+			if err := a.marshaler.Heartbeat(topic, partID, lastOffset); err != nil {
+				log.Errorf("[%s:%d] Admin failed to heartbeat", topic, partID)
+				log.Errorf("Admin is now unhealthy and will not reset offsets.")
+				atomic.StoreInt32(a.claimHealth, 0)
+				return
+			}
+			atomic.AddInt32(a.steps, 1)
+			time.Sleep(<-a.marshaler.cluster.jitters)
+		}
+	}
+}
+
+// claimAndHeartbeat attempts to claim a partition released by a paused consumer.
+// It heartbeats the previous offset.
+func (a *consumerGroupAdmin) claimAndHeartbeat(topic string,
+	partID int, newOffset int64, stopHeartbeat chan struct{}) bool {
+
+	// Get current offsets, which we will try to keep claimHealth.
+	offsets, err := a.marshaler.GetPartitionOffsets(topic, partID)
```

Comment:
You don't need the earliest/latest offsets, so why not call `GetLastPartitionClaim` directly and use the current offset from that?